src/OpenTelemetry.Contrib.Instrumentation.Wcf/TelemetryClientMessageInspector.cs

```diff
@@ -0,0 +1,132 @@
+﻿// <copyright file="TelemetryClientMessageInspector.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using System.ServiceModel;
+using System.ServiceModel.Channels;
+using System.ServiceModel.Dispatcher;
+using OpenTelemetry.Context.Propagation;
+using OpenTelemetry.Contrib.Instrumentation.Wcf.Implementation;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Contrib.Instrumentation.Wcf
+{
+    /// <summary>
+    /// An <see cref="IClientMessageInspector"/> implementation which adds telemetry to outgoing requests.
+    /// </summary>
+    public class TelemetryClientMessageInspector : IClientMessageInspector
+    {
+        /// <inheritdoc/>
+        public object BeforeSendRequest(ref Message request, IClientChannel channel)
+        {
+            try
+            {
+                if (WcfInstrumentationActivitySource.Options == null || WcfInstrumentationActivitySource.Options.OutgoingRequestFilter?.Invoke(request) == false)
+                {
+                    WcfInstrumentationEventSource.Log.RequestIsFilteredOut(request.Headers.Action);
+                    return new State
+                    {
+                        SuppressionScope = this.SuppressDownstreamInstrumentation(),
```

Comment:
@cijothomas @alanwest I ran into an interesting situation here with SuppressDownstreamInstrumentation. I think typically we only enter the suppression scope once Activity has been created. But in the case of a filtered request, we don't create activity and we don't enter suppression, therefore the downstream instrumentation always fires. I think the user's intention when filtering a request is to not have it show up in the trace so having the downstream instrumentation appear will be surprising. I fixed it by entering suppression even for filtered requests. Bringing it up here because it could be an issue elswhere, like gRPC instrumentation?

Reply:
Oh yea, this is interesting. gRPC instrumentation does not currently have a config for filtering requests. Though, I agree that it makes sense for downstream instrumentation to be suppressed in this scenario.